File: android/src/main/kotlin/twilio/flutter/twilio_conversations/PluginHandler.kt
package twilio.flutter.twilio_conversations

import android.content.Context
import androidx.annotation.NonNull
import android.util.Log
import com.twilio.conversations.CallbackListener
import com.twilio.conversations.ConversationsClient
import com.twilio.util.ErrorInfo
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import twilio.flutter.twilio_conversations.listeners.ChatListener
import twilio.flutter.twilio_conversations.methods.ChannelMethods
import twilio.flutter.twilio_conversations.methods.ChannelsMethods
import twilio.flutter.twilio_conversations.methods.ChatClientMethods
import twilio.flutter.twilio_conversations.methods.MemberMethods
import twilio.flutter.twilio_conversations.methods.MembersMethods
import twilio.flutter.twilio_conversations.methods.MessageMethods
import twilio.flutter.twilio_conversations.methods.MessagesMethods
import twilio.flutter.twilio_conversations.methods.UserMethods
import twilio.flutter.twilio_conversations.methods.UsersMethods

class PluginHandler(private val pluginInstance: TwilioConversationsPlugin, private val applicationContext: Context) : MethodCallHandler {
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        Log.d("TwilioInfo", "TwilioConversationsPlugin.onMethodCall => received ${call.method}")
        if (call.method != "cancel" && call.method != "listen") {
            when (call.method) {
                "debug" -> debug(call, result)
                "create" -> create(call, result)
                "registerForNotification" -> pluginInstance.registerForNotification(call, result)
                "unregisterForNotification" -> pluginInstance.unregisterForNotification(call, result)
                "handleReceivedNotification" -> pluginInstance.handleReceivedNotification(call, result)

                "ChatClient#updateToken" -> ChatClientMethods.updateToken(call, result)
                "ChatClient#shutdown" -> ChatClientMethods.shutdown(call, result)

                "User#unsubscribe" -> UserMethods.unsubscribe(call, result)

                "Users#getChannelUserDescriptors" -> UsersMethods.getChannelUserDescriptors(call, result)
                "Users#getUserDescriptor" -> UsersMethods.getUserDescriptor(call, result)
                "Users#getAndSubscribeUser" -> UsersMethods.getAndSubscribeUser(call, result)

                "Channel#join" -> ChannelMethods.join(call, result)
                "Channel#leave" -> ChannelMethods.leave(call, result)
                "Channel#typing" -> ChannelMethods.typing(call, result)
                "Channel#destroy" -> ChannelMethods.destroy(call, result)
                "Channel#getMessagesCount" -> ChannelMethods.getMessagesCount(call, result)
                "Channel#getUnreadMessagesCount" -> ChannelMethods.getUnreadMessagesCount(call, result)
                "Channel#getMembersCount" -> ChannelMethods.getMembersCount(call, result)
                "Channel#setAttributes" -> ChannelMethods.setAttributes(call, result)
                "Channel#getFriendlyName" -> ChannelMethods.getFriendlyName(call, result)
                "Channel#setFriendlyName" -> ChannelMethods.setFriendlyName(call, result)
                "Channel#getNotificationLevel" -> ChannelMethods.getNotificationLevel(call, result)
                "Channel#setNotificationLevel" -> ChannelMethods.setNotificationLevel(call, result)
                "Channel#getUniqueName" -> ChannelMethods.getUniqueName(call, result)
                "Channel#setUniqueName" -> ChannelMethods.setUniqueName(call, result)

                "Channels#getChannel" -> ChannelsMethods.getChannel(pluginInstance, call, result)
                "Channels#getPublicChannelsList" -> ChannelsMethods.getPublicChannelsList(call, result)
                "Channels#getUserChannelsList" -> ChannelsMethods.getUserChannelsList(call, result)
                "Channels#createChannel" -> ChannelsMethods.createChannel(pluginInstance, call, result)

                "Member#getUserDescriptor" -> MemberMethods.getUserDescriptor(call, result)
                "Member#getAndSubscribeUser" -> MemberMethods.getAndSubscribeUser(call, result)
                "Member#setAttributes" -> MemberMethods.setAttributes(call, result)

                "Members#getMembersList" -> MembersMethods.getMembersList(call, result)
                "Members#getMember" -> MembersMethods.getMember(call, result)
                "Members#addByIdentity" -> MembersMethods.addByIdentity(call, result)
                "Members#inviteByIdentity" -> MembersMethods.inviteByIdentity(call, result)
                "Members#removeByIdentity" -> MembersMethods.removeByIdentity(call, result)

                "Message#updateMessageBody" -> MessageMethods.updateMessageBody(call, result)
                "Message#setAttributes" -> MessageMethods.setAttributes(call, result)
                "Message#getMedia" -> MessageMethods.getMedia(call, result)

                "Messages#sendMessage" -> MessagesMethods.sendMessage(pluginInstance, call, result)
                "Messages#removeMessage" -> MessagesMethods.removeMessage(call, result)
                "Messages#getMessagesBefore" -> MessagesMethods.getMessagesBefore(call, result)
                "Messages#getMessagesAfter" -> MessagesMethods.getMessagesAfter(call, result)
                "Messages#getLastMessages" -> MessagesMethods.getLastMessages(call, result)
                "Messages#getMessageByIndex" -> MessagesMethods.getMessageByIndex(call, result)
                "Messages#setLastReadMessageIndexWithResult" -> MessagesMethods.setLastReadMessageIndexWithResult(call, result)
                "Messages#advanceLastReadMessageIndexWithResult" -> MessagesMethods.advanceLastReadMessageIndexWithResult(call, result)
                "Messages#setAllMessagesReadWithResult" -> MessagesMethods.setAllMessagesReadWithResult(call, result)
                "Messages#setNoMessagesReadWithResult" -> MessagesMethods.setNoMessagesReadWithResult(call, result)

                else -> result.notImplemented()
            }
        }
    }

    private fun create(call: MethodCall, result: MethodChannel.Result) {
        Log.d("TwilioInfo", "TwilioConversationsPlugin.create => called")

        val token = call.argument<String>("token")
        val propertiesObj = call.argument<Map<String, Any>>("properties")
        if (token == null) {
            return result.error("ERROR", "Missing token", null)
        }

        if (propertiesObj == null) {
            return result.error("ERROR", "Missing properties", null)
        }

        val callRegion = propertiesObj["region"] as String?
        val callDeferCA = propertiesObj["deferCA"] as Boolean?


        val currentChatClient = TwilioConversationsPlugin.chatClient
        val currentChatClientRegion = TwilioConversationsPlugin.chatClientRegion
        val currentChatClientDeferCA = TwilioConversationsPlugin.chatClientDeferCA

        try {
            if (currentChatClient == null) {
                Log.d("TwilioInfo", "TwilioConversationsPlugin.create => making a fresh ChatClient")
            } else {
                Log.w("TwilioInfo", "TwilioConversationsPlugin.create => ChatClient is already exists.")
                if (callRegion != currentChatClientRegion) {
                    result.error("ERROR", "ChatClient already exists with a different region", null)
                    return
                } else if (callDeferCA != currentChatClientDeferCA) {
                    result.error("ERROR", "ChatClient already exists with a different deferCA", null)
                    return
                }
            }

            val propertiesBuilder = ConversationsClient.Properties.newBuilder()
            if (callRegion != null) {
                Log.d("TwilioInfo", "TwilioConversationsPlugin.create => setting Properties.region to '${callRegion}'")
                propertiesBuilder.setRegion(callRegion)
            }

            if (callDeferCA != null) {
                Log.d("TwilioInfo", "TwilioConversationsPlugin.create => setting Properties.setDeferCertificateTrustToPlatform to '${callDeferCA}'")
                propertiesBuilder.setDeferCertificateTrustToPlatform(callDeferCA)
            }

            pluginInstance.chatListener = ChatListener(pluginInstance, propertiesBuilder.createProperties())

            if (currentChatClient != null) {
                val chatClientMap = Mapper.chatClientToMap(pluginInstance, currentChatClient)
                result.success(chatClientMap)
                return
            }

            ConversationsClient.create(applicationContext, token, pluginInstance.chatListener.properties, object : CallbackListener<ConversationsClient> {
                override fun onSuccess(chatClient: ConversationsClient) {
                    Log.d("Twilio init success", "TwilioConversationsPlugin.create => ChatClient.create onSuccess: myIdentity is '${chatClient.myIdentity}'")
                    try {
                        TwilioConversationsPlugin.chatClient = chatClient
                        TwilioConversationsPlugin.chatClientRegion = callRegion
                        TwilioConversationsPlugin.chatClientDeferCA = callDeferCA

                        val chatClientMap = Mapper.chatClientToMap(pluginInstance, chatClient)
                        result.success(chatClientMap)
                    } catch (e: Exception) {
                        Log.d("TwilioInfo", "TwilioConversationsPlugin.create => ChatClient.create onSuccess: failed to give result")
                        Log.d("Twilio error", "TwilioConversationsPlugin.create => ChatClient.create onSuccess: failed to give result: $e : $chatClient : ${Mapper.chatClientToMap(pluginInstance, chatClient)}")
                        result.error("$e", "Failed to map chatClient", null)
                    }
                }

                override fun onError(errorInfo: ErrorInfo) {
                        try {
                            with(pluginInstance) { debug("TwilioConversationsPlugin.create => ChatClient.create onError: $errorInfo") }
                            result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
                        } catch (e1: Exception) {
                            Log.d("TwilioInfo", "TwilioConversationsPlugin.create => ChatClient.create error: failed to send result of onError: $e1")
                        }
                }
            })
        } catch (e: Exception) {
            result.error("ERROR", e.toString(), e)
        }
    }

    private fun debug(call: MethodCall, result: MethodChannel.Result) {
        val enableNative = call.argument<Boolean>("native")
        val enableSdk = call.argument<Boolean>("sdk")

        if (enableSdk != null && enableSdk) {
            ConversationsClient.setLogLevel(ConversationsClient.LogLevel.DEBUG)
        }

        if (enableNative != null) {
            pluginInstance.nativeDebug = enableNative
            result.success(enableNative)
        } else {
            result.error("MISSING_PARAMS", "Missing 'native' parameter", null)
        }
    }
}

File: flutter_twilio_conversations/ios/Classes/PluginHandler.swift
import Flutter
import Foundation
import TwilioConversationsClient

public class PluginHandler {
    // swiftlint:disable:next function_body_length
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        SwiftTwilioConversationsPlugin.debug("PluginHandler.handle => received \(call.method)")
        switch call.method {
        case "debug":
            debug(call, result: result)
        case "create":
            create(call, result: result)
        case "registerForNotification":
            SwiftTwilioConversationsPlugin.instance?.registerForNotification(call, flutterResult: result)
        case "unregisterForNotification":
            SwiftTwilioConversationsPlugin.instance?.unregisterForNotification(call, flutterResult: result)
        case "handleReceivedNotification":
            SwiftTwilioConversationsPlugin.instance?.handleReceivedNotification(call, flutterResult: result)
        case "ChatClient#updateToken":
            ChatClientMethods.updateToken(call, result: result)
        case "ChatClient#shutdown":
            ChatClientMethods.shutdown(call, result: result)

        case "User#unsubscribe":
            UserMethods.unsubscribe(call, result: result)

        case "Users#getChannelUserDescriptors":
            UsersMethods.getChannelUserDescriptors(call, result: result)
        case "Users#getUserDescriptor":
            UsersMethods.getUserDescriptor(call, result: result)
        case "Users#getAndSubscribeUser":
            UsersMethods.getAndSubscribeUser(call, result: result)

        case "Channel#join":
            ChannelMethods.join(call, result: result)
        case "Channel#leave":
            ChannelMethods.leave(call, result: result)
        case "Channel#typing":
            ChannelMethods.typing(call, result: result)
        case "Channel#destroy":
            ChannelMethods.destroy(call, result: result)
        case "Channel#getMessagesCount":
            ChannelMethods.getMessagesCount(call, result: result)
        case "Channel#getUnreadMessagesCount":
            ChannelMethods.getUnreadMessagesCount(call, result: result)
        case "Channel#getMembersCount":
            ChannelMethods.getMembersCount(call, result: result)
        case "Channel#setAttributes":
            ChannelMethods.setAttributes(call, result: result)
        case "Channel#getFriendlyName":
            ChannelMethods.getFriendlyName(call, result: result)
        case "Channel#setFriendlyName":
            ChannelMethods.setFriendlyName(call, result: result)
        case "Channel#getNotificationLevel":
            ChannelMethods.getNotificationLevel(call, result: result)
        case "Channel#setNotificationLevel":
            ChannelMethods.setNotificationLevel(call, result: result)
        case "Channel#getUniqueName":
            ChannelMethods.getUniqueName(call, result: result)
        case "Channel#setUniqueName":
            ChannelMethods.setUniqueName(call, result: result)

        case "Channels#createChannel":
            ChannelsMethods.createChannel(call, result: result)
        case "Channels#getChannel":
            ChannelsMethods.getChannel(call, result: result)
        case "Channels#getUserChannelsList":
            ChannelsMethods.getUserChannelsList(call, result: result)
        case "Channels#getMembersByIdentity":
            ChannelsMethods.getMembersByIdentity(call, result: result)

        case "Member#getUserDescriptor":
            MemberMethods.getUserDescriptor(call, result: result)
        case "Member#getAndSubscribeUser":
            MemberMethods.getAndSubscribeUser(call, result: result)
        case "Member#setAttributes":
            MemberMethods.setAttributes(call, result: result)

        case "Members#getMembersList":
            MembersMethods.getMembersList(call, result: result)
        case "Members#getMember":
            MembersMethods.getMember(call, result: result)
        case "Members#addByIdentity":
            MembersMethods.addByIdentity(call, result: result)
        case "Members#inviteByIdentity":
            MembersMethods.inviteByIdentity(call, result: result)
        case "Members#removeByIdentity":
            MembersMethods.removeByIdentity(call, result: result)

        case "Message#updateMessageBody":
            MessageMethods.updateMessageBody(call, result: result)
        case "Message#setAttributes":
            MessageMethods.setAttributes(call, result: result)
        case "Message#getMedia":
            MessageMethods.getMedia(call, result: result)

        case "Messages#sendMessage":
            MessagesMethods.sendMessage(call, result: result)
        case "Messages#removeMessage":
            MessagesMethods.removeMessage(call, result: result)
        case "Messages#getMessagesBefore":
            MessagesMethods.getMessagesBefore(call, result: result)
        case "Messages#getMessagesAfter":
            MessagesMethods.getMessagesAfter(call, result: result)
        case "Messages#getLastMessages":
            MessagesMethods.getLastMessages(call, result: result)
        case "Messages#getMessageByIndex":
            MessagesMethods.getMessageByIndex(call, result: result)
        case "Messages#setLastReadMessageIndexWithResult":
            MessagesMethods.setLastReadMessageIndexWithResult(call, result: result)
        case "Messages#advanceLastReadMessageIndexWithResult":
            MessagesMethods.advanceLastReadMessageIndexWithResult(call, result: result)
        case "Messages#setAllMessagesReadWithResult":
            MessagesMethods.setAllMessagesReadWithResult(call, result: result)
        case "Messages#setNoMessagesReadWithResult":
            MessagesMethods.setNoMessagesReadWithResult(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func debug(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let enableNative = arguments["native"] as? Bool else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'native' parameter", details: nil))
        }

        guard let enableSdk = arguments["sdk"] as? Bool else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'sdk' parameter", details: nil))
        }

        SwiftTwilioConversationsPlugin.nativeDebug = enableNative
        if enableSdk {
            TwilioConversationsClient.setLogLevel(TCHLogLevel.debug)
        }
        result(enableNative)
    }

    private func create(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        SwiftTwilioConversationsPlugin.debug("TwilioConversationsPlugin.create => called")

        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let token = arguments["token"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'token' parameter", details: nil))
        }

        guard let propertiesObj = arguments["properties"] as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'token' parameter", details: nil))
        }

        let properties = TwilioConversationsClientProperties()
        if let region = propertiesObj["region"] as? String {
            properties.region = region
        }

        let flutterResult = result

        SwiftTwilioConversationsPlugin.chatListener = ChatListener(token, properties)

        TwilioConversationsClient.conversationsClient(withToken: token, properties: properties, delegate: SwiftTwilioConversationsPlugin.chatListener, completion: {(result: TCHResult, chatClient: TwilioConversationsClient?) -> Void in
            if result.isSuccessful {
                SwiftTwilioConversationsPlugin.debug("TwilioConversationsPlugin.create => ChatClient.create onSuccess: myIdentity is '\(chatClient?.user?.identity ?? "unknown")'")
                SwiftTwilioConversationsPlugin.chatListener?.chatClient = chatClient
                flutterResult(Mapper.chatClientToDict(chatClient))
            } else {
                SwiftTwilioConversationsPlugin.debug("TwilioConversationsPlugin.create => ChatClient.create onError: \(String(describing: result.error))")
            }
            } as TCHTwilioClientCompletion)
    }
}
